### PR TITLE
Fix reconsensus

### PR DESCRIPTION
**Pull request by mmolari**
Fixed a bug in reconsensus: when transferring a substitution to the consensus, sites that are deleted should not receive the substitution.

This caused a failure in node removal downstream, in which empty nodes were not removed because they featured a substitution, which was erroneously located on a deleted site.

I added additional debug sanity checks and fixed the bug. I also added the edge-case to the test set.

### CODE REVIEW

File: packages/pangraph/src/pangraph/pangraph_node.rs [hunk 1]
```diff
@@ -41,12 +41,12 @@ impl PangraphNode {
     }
   }
 
-  pub fn len(&self) -> usize {
```
Comment (mmolari) [inline packages/pangraph/src/pangraph/pangraph_node.rs:44]: I this this function had a bug. But due to possibly having circular genomes, the length of a node cannot always be determined by only start-end positions

File: packages/pangraph/src/pangraph/pangraph_node.rs [hunk 2]
```diff
@@ -41,12 +41,12 @@ impl PangraphNode {
...
-    self.position.1.saturating_sub(self.position.1)
-  }
-
-  pub fn is_empty(&self) -> bool {
```
Comment (mmolari) [inline packages/pangraph/src/pangraph/pangraph_node.rs:48]: Therefore also this function was wrong. I thought about defining `is_emtpy` as checking if the start position coincides with the end position, but this is also wrong in one edge case: when the path is circular and composed of a single node, then the start/end position of the node can be the same even if the node is not empty.

File: packages/pangraph/src/reconsensus/reconsensus.rs [hunk 3]
```diff
@@ -87,8 +91,11 @@ fn reconsensus_mutations(block: &mut PangraphBlock) -> Result<(), Report> {
       let subs_at_pos: Vec<_> = edit.subs.iter().filter(|s| s.pos == pos).cloned().collect();
       match subs_at_pos.len() {
         0 => {
-          edit.subs.push(Sub::new(pos, original));
```
Comment (mmolari) [inline packages/pangraph/src/reconsensus/reconsensus.rs:90]: This fixed the bug: the substitution should be applied only if the position is not deleted.

File: packages/pangraph/src/reconsensus/reconsensus.rs [hunk 4]
```diff
@@ -203,14 +226,14 @@ mod tests {
     //               01234567890123456789012
     //    node 1)    .T...--..........A.....
     //    node 2)    .T...--...C......|.....
-    //    node 3)    .T...--...C............
```
Comment (mmolari) [inline packages/pangraph/src/reconsensus/reconsensus.rs:206]: changed the test so that it hits this edge-case

File: packages/pangraph/src/reconsensus/remove_nodes.rs [hunk 5]
```diff
@@ -47,10 +74,19 @@ fn remove_nodes_from_graph(graph: &mut Pangraph, node_ids: &[NodeId]) {
...
+    // let path_nodes = &mut graph.paths.get_mut(&path_id).unwrap().nodes;
+    // let node_idx = path_nodes.iter().position(|&n| n == node_id).unwrap();
+    // path_nodes.remove(node_idx);
+
+    //  remove all nodes with the same id
     let path_nodes = &mut graph.paths.get_mut(&path_id).unwrap().nodes;
-    let node_idx = path_nodes.iter().position(|&n| n == node_id).unwrap();
-    path_nodes.remove(node_idx);
+    // while there is a node with this id, remove it:
+    while let Some(node_idx) = path_nodes.iter().position(|&n| n == node_id) {
```
Comment (mmolari) [inline packages/pangraph/src/reconsensus/remove_nodes.rs:87]: This is the second fix: sometimes a path could have more than one empty node with the same id. The same id happens only if the nodes are adjacent. This removes all of them.

A more elegant solution might be to not produce these empty nodes at all (modify the merging procedure?) or to make sure they are generated with different id.
